sdk/data/azcosmos/cosmos_location_cache.go

```diff
@@ -0,0 +1,303 @@
+	lc.mapMutex.RLock()
+	defer lc.mapMutex.RUnlock()
+	if time.Since(lc.lastUpdateTime) > lc.unavailableLocationExpirationTime && len(lc.locationUnavailabilityInfoMap) > 0 {
+		err := lc.update(nil, nil, nil, lc.enableMultipleWriteLocations)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return lc.locationInfo.readEndpts, nil
+}
+
+func (lc *locationCache) writeEndpts() ([]url.URL, error) {
+	lc.mapMutex.RLock()
+	defer lc.mapMutex.RUnlock()
+	if time.Since(lc.lastUpdateTime) > lc.unavailableLocationExpirationTime && len(lc.locationUnavailabilityInfoMap) > 0 {
+		err := lc.update(nil, nil, nil, lc.enableMultipleWriteLocations)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return lc.locationInfo.writeEndpts, nil
+}
+
+func (lc *locationCache) getLocation(endpoint url.URL) string {
+	firstLoc := ""
+	lc.rwMutex.RLock()
+	defer lc.rwMutex.RUnlock()
+	for location, uri := range lc.locationInfo.availWriteEndptsByLocation {
+		if uri == endpoint {
+			return location
+		}
+		if firstLoc == "" {
+			firstLoc = location
+		}
+	}
+
+	for location, uri := range lc.locationInfo.availReadEndptsByLocation {
+		if uri == endpoint {
+			return location
+		}
+	}
+
+	if endpoint == lc.defaultEndpt && !lc.canUseMultipleWriteLocs() {
+		if len(lc.locationInfo.availWriteEndptsByLocation) > 0 {
+			return firstLoc
+		}
+	}
+	return ""
+}
+
+func (lc *locationCache) canUseMultipleWriteLocs() bool {
+	lc.rwMutex.RLock()
+	defer lc.rwMutex.RUnlock()
+	return lc.useMultipleWriteLocations && lc.enableMultipleWriteLocations
+}
+
+func (lc *locationCache) markEndptUnavailableForRead(endpoint url.URL) error {
+	return lc.markEndptUnavailable(endpoint, read)
+}
+
+func (lc *locationCache) markEndptUnavailableForWrite(endpoint url.URL) error {
+	return lc.markEndptUnavailable(endpoint, write)
+}
+
+func (lc *locationCache) markEndptUnavailable(endpoint url.URL, op opType) error {
+	now := time.Now()
+	lc.mapMutex.Lock()
+	if info, ok := lc.locationUnavailabilityInfoMap[endpoint]; ok {
+		info.lastCheckTime = now
+		info.unavailableOps |= op
+		lc.locationUnavailabilityInfoMap[endpoint] = info
+	} else {
+		info = locationUnavailabilityInfo{
+			lastCheckTime:  now,
+			unavailableOps: op,
+		}
+		lc.locationUnavailabilityInfoMap[endpoint] = info
+	}
+	lc.mapMutex.Unlock()
+	err := lc.update(nil, nil, nil, lc.enableMultipleWriteLocations)
+	return err
+}
+
+func (lc *locationCache) dbAcctRead(dbAcct AcctProperties) error {
+	return lc.update(dbAcct.writeRegions, dbAcct.readRegions, nil, dbAcct.enableMultipleWriteLocations)
+}
+
+func (lc *locationCache) refreshStaleEndpts() {
+	lc.mapMutex.Lock()
+	for endpoint, info := range lc.locationUnavailabilityInfoMap {
+		t := time.Since(info.lastCheckTime)
+		if t > lc.unavailableLocationExpirationTime {
+			delete(lc.locationUnavailabilityInfoMap, endpoint)
+		}
+	}
+	lc.mapMutex.Unlock()
+}
+
+func (lc *locationCache) isEndptUnavailable(endpoint url.URL, ops opType) bool {
```

Comment:
We already have OperationType enum available, add any you need there https://github.com/Azure/azure-sdk-for-go/blob/main/sdk/data/azcosmos/operation_type.go

Reply:
I think the type that I defined is still needed, perhaps a better name for it would be something along the lines of `requestedOperations`? Since these are treated like flags for the location cache to indicate one or more operations to check availability for.